sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala

```diff
@@ -676,6 +678,43 @@ class KeyValueGroupedDataset[K, V] private[sql](
     )
   }
 
+  private[sql] def transformWithState[U: Encoder](
+       statefulProcessor: StatefulProcessor[K, V, U],
+       timeoutMode: TimeoutMode,
+       eventTimeColumnName: String,
+       outputMode: OutputMode): Dataset[U] = {
+    val existingWatermarkDelay = logicalPlan.flatMap {
+      case EventTimeWatermark(_, delay, _) => Seq(delay)
+      case _ => Seq()
+    }
+
+    if (existingWatermarkDelay.isEmpty) {
+      throw QueryCompilationErrors.cannotAssignEventTimeColumn()
```

Comment:
Do we need additional checks in the `UnsupportedOperationChecker` too ?

Reply:
I dont think we do. `UnsupportedOperationChecker` will throw an error if there are stateful operators chained after transformWithState, but user does not provide the eventTimeColumnName in tws.